GH-4155 - Increase gRPC message size

komalali opened the pull request:
Fixes: #4155 

Related: https://github.com/pulumi/pulumi-policy/pull/239

Comments:
* I tried to minimize the number of individual places where that `1024 * 1024 * 400` number is referenced by creating a util function, but that only helps for the Go code. Is there a preferred way of adding such a widespread setting?
* Is the dotnet code adequate? Haven't ever worked in that lang before.
* For #4155 to truly be fixed, providers will need to be regenerated with pulumi-terraform-bridge that pulls this updated code and the related pulumi-policy PR must be merged.

lukehoban commented on sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs:
```diff
@@ -12,7 +13,9 @@ internal class GrpcEngine : IEngine
         public GrpcEngine(string engine)
         {
-            this._engine = new Engine.EngineClient(new Channel(engine, ChannelCredentials.Insecure));
+            var maxRpcMessageSize = 400 * 1024 * 1024;
```
Perhaps add a comment explaining this constant?  Here and in all the other similar locations.

lukehoban commented on sdk/go/pulumi/context.go:
```diff
@@ -27,13 +27,13 @@ import (
 	"github.com/pulumi/pulumi/sdk/v2/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/rpcutil"
 	pulumirpc "github.com/pulumi/pulumi/sdk/v2/proto/go"
+	"google.golang.org/grpc"
```
Very minor - but we've generally maintained three blocks of imports (1) builtins (2) external dependencies (3) internal dependencies.

komalali replied:
That's what those 3 blocks represent! Okay good to know. Will fix.

lukehoban commented on sdk/nodejs/runtime/settings.ts:
```diff
@@ -25,6 +26,9 @@ const resrpc = require("../proto/resource_grpc_pb.js");
 const structproto = require("google-protobuf/google/protobuf/struct_pb.js");
 
+const maxRPCMessageSize: number = 1024 * 1024 * 400;
+const grpcChannelOptions = <ChannelOptions>{"grpc.max_receive_message_length": maxRPCMessageSize};
```
Is the cast needed?  Or is this an option?

```suggestion
const grpcChannelOptions: ChannelOptions = {"grpc.max_receive_message_length": maxRPCMessageSize};
```

komalali replied:
ah yup totally, typescript n00b moment.

lukehoban commented on sdk/nodejs/tests/runtime/langhost/run.spec.ts:
```diff
@@ -1136,6 +1136,24 @@ describe("rpc", () => {
+                    },
+                };
+            },
+        },
```
Do these tests actually go through the RPC layer?  I don't recall off the top of my head - but I feel like they do not?  cc @pgavlin 

One place we could add tests to be sure they hit the RPC is in `/test/integration`.  Those will run the full "normal" CLI/Engine/LangHost setup.

lukehoban replied:
BTW - same likely applies to other languages.

komalali replied:
@lukehoban yes they do. I made sure by checking that they fail on master with the same resource exhausted error. I can definitely add integration tests too but I am certain that these tests go through the langhost <-> engine RPC layer.

komalali commented:
Okay! I added in the integration tests here, added the skipped Node.js test in Windows to the list in https://github.com/pulumi/pulumi/issues/3811 and I think we're good to go here.
